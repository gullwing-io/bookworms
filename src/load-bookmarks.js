import got from 'got';
import yaml from 'js-yaml';
import { readFileSync } from 'fs';

const fetchBookmarkConfig = async (path, asYAML = false) => {
    let type;
    let body;
    if (shouldFetchFromLocal(path)) {
        type = 'local';
        if (!asYAML) {
            body = returnResponseAsObject(fetchLocaleBookmarkYAML(path));
        } else {
            body = fetchLocaleBookmarkYAML(path);
        }
    } else {
        type = 'remote';
        if (!asYAML) {
            const response = await fetchRemoteBoomarkYMAL(path)
            body = returnResponseAsObject(response.body, response.path)
        } else {
            const reponse =  await fetchRemoteBoomarkYMAL(path)
            body = reponse.body
        }
    }
    return {
        type,
        body
    }
}

const shouldFetchFromLocal = (path) => {
    try {
        new URL (path);
        return false;   
    } catch (e) {
        return true;
    }
}

const fetchLocaleBookmarkYAML = (path) => {
    try {
        return readFileSync(path, 'utf8');
    } catch(error) {
        // todo add tests around this error handling
        console.error(`${path} returned ${error.message}`)
        process.exit(1)
    }
}

const fetchRemoteBoomarkYMAL = async (path) =>{
    try {
        const { body, statusCode } = await got(path);
        return returnResponseOn2xx(body, statusCode, path);
    } catch (error) {
        // todo add tests around this error handling
        console.error(`${path} returned ${error.message}`)
        process.exit(1)
    }
}

const returnResponseOn2xx = (body, statusCode, path) => {
    if (statusCode >= 200 && statusCode < 300) {
        return {body, path}
    } else {
        throw new Error(`${path} returned ${statusCode}`)
    }
}

const returnResponseAsObject = (body, path) => {
    if (typeof body === 'object') {
        return body
    } else {
        const json = yaml.load(body);
        // this is bit gross but its what the lib returns
        if (json === "undefined") {
            try {
                return JSON.parse(response);
            } catch(e) {
                throw new Error(`${path} returned invalid bookworms bookmarks structure`)
            }
        } else {
            return json
        }
    }
}

export {fetchBookmarkConfig, shouldFetchFromLocal, fetchLocaleBookmarkYAML, fetchRemoteBoomarkYMAL, returnResponseOn2xx, returnResponseAsObject}
