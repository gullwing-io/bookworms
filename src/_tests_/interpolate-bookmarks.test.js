import {fetchBookmarkConfig} from '../load-bookmarks';
import interpolateBookmarks from '../interpolate-bookmarks';

describe('interpolateBookmarks', () => {
    test('should correctly return yaml without placeholders', async () => {
        const {body} = await fetchBookmarkConfig('./demo/config/bookmarks.yaml', true);
        const interpolatedYaml = await interpolateBookmarks(body);
        expect(interpolatedYaml).toEqual(body);
    })
    test('should correctly return yaml with placeholders', async () => {
        const {body} = await fetchBookmarkConfig('./demo/config/interpolate-bookmark.template.yaml', true);
        const interpolatedYaml = await interpolateBookmarks(body);
        
        expect(interpolatedYaml).toMatchSnapshot();
    })
})