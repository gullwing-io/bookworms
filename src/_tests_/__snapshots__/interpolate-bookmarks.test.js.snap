// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`interpolateBookmarks should correctly return yaml with placeholders 1`] = `
"label: Interpolated bookmarks
description: This is an example of interpolating YAML files
folders:
  -
    label: Bookworms
    description: These are sample bookmarks to teach you how Bookworms works
    folders:
      - label: folder 1
        description: This is to describe the folder structure
        folders:
          - label: sub folder 1
            description: This is to describe the sub folder structure
            bookmarks:
              - label: sample url 1
                description: this is used to describe the bookmark
                href: https://www.mywebsite.com
      - label: folder 2
        folders:
          - label: sub folder 2
            bookmarks:
              - label: sample url 2
                description: this is used to describe the bookmark
                href: https://www.mywebsite.com
              - label: sample url 3
                description: this is used to describe the bookmark
                href: https://www.mywebsite.com
          - label: sub folder 3
            bookmarks:
              - label: sample url 4
                description: this is used to describe the bookmark
                href: https://www.mywebsite.com
              - label: sample url 5
                description: this is used to describe the bookmark
                href: https://www.mywebsite.com
  -
    # This config file is used to show how you can merge bookmark config
    label: Bookworms for merging
    description: These are sample bookmarks to merge with bookmarks.yaml
    folders:
      - label: folder 3
        description: This is to describe the folder structure
        folders:
          - label: sub folder 1
            description: This is to describe the sub folder structure
            bookmarks:
              - label: sample url 1
                description: this is used to describe the bookmark
                href: https://www.mywebsite.com
    "
`;
